src/nxt_otel.c

```diff
@@ -0,0 +1,344 @@
+    nxt_otel_send_trace(r->otel->trace);
+    r->otel->trace = NULL;
+}
+
+static inline void
+nxt_otel_span_collect(nxt_task_t *t, nxt_http_request_t *r)
+{
+    nxt_log(t, NXT_LOG_DEBUG, "collecting span by adding the task to the fast work queue");
+    nxt_work_queue_add(&t->thread->engine->fast_work_queue,
+                       nxt_otel_send_trace_and_span_data, t, r, NULL);
+    nxt_otel_state_transition(r->otel, 0);
+}
+
+static void
+nxt_otel_error(nxt_task_t *t, nxt_http_request_t *r)
+{
+    // purposefully not using state transition helper
+    r->otel->status = 0;
+    nxt_log(t, NXT_LOG_ERR, "otel error condition");
+    // if r->otel->trace it WILL leak here.
+    // TODO Phase 2: drop trace without sending it somehow?
+}
+
+nxt_int_t
+nxt_otel_parse_traceparent(void *ctx, nxt_http_field_t *field, uintptr_t data)
+{
+    nxt_http_request_t *r;
+    char               *copy;
+
+    /* For information on parsing the traceparent header:
+     * https://www.w3.org/TR/trace-context/#traceparent-header
+     * A summary of the traceparent header value format follows:
+     * Traceparent: "$a-$b-$c-$d"
+     *   a. version (2 hex digits) (ff is forbidden)
+     *   b. trace_id (32 hex digits) (all zeroes forbidden)
+     *   c. parent_id (16 hex digits) (all zeroes forbidden)
+     *   d. flags (2 hex digits)
+     */
+
+    r = ctx;
+    if (field->value_length != NXT_OTEL_TRACEPARENT_LEN) {
+        goto error_state;
+    }
+
+    /* strsep is destructive so we make a copy of the field
+     */
+    copy = nxt_mp_zalloc(r->mem_pool, field->value_length+1);
+    if (copy == NULL) {
+        goto error_state;
+    }
+    memcpy(copy, field->value, field->value_length);
+
+    r->otel->version = (u_char *) strsep(&copy, "-");
+    r->otel->trace_id = (u_char *) strsep(&copy, "-");
+    r->otel->parent_id = (u_char *) strsep(&copy, "-");
+    r->otel->trace_flags = (u_char *) strsep(&copy, "-");
+
+    if (!r->otel->version ||
+        !r->otel->trace_id ||
+        !r->otel->parent_id ||
+        !r->otel->trace_flags) {
+        goto error_state;
+    }
+
+    return NXT_OK;
+
+ error_state:
+    nxt_otel_state_transition(r->otel, NXT_OTEL_ERROR_STATE);
+    return NXT_ERROR;
+}
+
+nxt_int_t
+nxt_otel_parse_tracestate(void *ctx, nxt_http_field_t *field, uintptr_t data)
+{
+    nxt_http_request_t *r;
+    nxt_str_t     s;
+    nxt_http_field_t *f;
+
+    s.length = field->value_length;
+    s.start = field->value;
+    r = ctx;
+    r->otel->trace_state = s;
+
+    // maybe someday this should get sent down into the otel lib
+
+    f = nxt_list_add(r->resp.fields);
+    if (f) {
+      *f = *field;
+    }
+
+    return NXT_OK;
+}
+
+inline nxt_int_t
+nxt_otel_validate_endpoint(nxt_conf_validation_t *vldt, nxt_conf_value_t *value, void *data)
+{
+    // This function is a stub for now
+    return NXT_OK;
+}
+
+nxt_int_t
+nxt_otel_validate_batch_size(nxt_conf_validation_t *vldt, nxt_conf_value_t *value, void *data)
```

Comment:
Watch your line lengths. We're pretty strict about not going over 80 characters...